### ISSUE TITLE
React-apollo 3.0 with aws-appsync-react

### ISSUE THREAD

**Issue by Xocix** [awslabs/aws-mobile-appsync-sdk-js#448]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Installing react-apollo 3.0 makes the Rehydrated component stop working

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
Error message recieved: The context `client` is marked as required in `Rehydrated`, but its value is `undefined`.

**What is the expected behavior?**
Rehydrated should have a client to be able to rehydrate

**Comment by noincomedev** [comment]
I am having the same issue!

**Comment by Xocix** [comment]
My workaround is to create my own Rehydrated component that takes in client as a prop. 
Also i needed to add a resolution to the package.json otherwise it was crashing for another error.

 `    "resolutions": {
        "apollo-client": "2.6.3"
    }`

```
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import AWSAppSyncClient from 'aws-appsync';

const Rehydrate = props => (
    <div
        className={`awsappsync ${
            props.rehydrated
                ? 'awsappsync--rehydrated'
                : 'awsappsync--rehydrating'
        }`}
    >
        {props.rehydrated ? props.children : <span>Loading...</span>}
    </div>
);

class Rehydrated extends Component {
    static propTypes = {
        render: PropTypes.func,
        children: PropTypes.node,
        loading: PropTypes.node,
        client: PropTypes.instanceOf(AWSAppSyncClient).isRequired,
    };

    constructor(props) {
        super(props);

        this.state = {
            rehydrated: false,
        };
    }

    async componentWillMount() {
        await this.props.client.hydrated();

        this.setState({
            rehydrated: true,
        });
    }

    render() {
        const { render, children, loading } = this.props;
        const { rehydrated } = this.state;

        if (render) return render({ rehydrated });

        if (children) {
            if (loading) return rehydrated ? children : loading;

            return <Rehydrate rehydrated={rehydrated}>{children}</Rehydrate>;
        }
    }
}

export default Rehydrated;```

**Comment by laurib** [comment]
This is my custom Rehydrated component. 

```
import React, { useContext, useEffect, useState } from 'react';
import { getApolloContext } from 'react-apollo';
import AWSAppSyncClient from 'aws-appsync';

const Rehydrated = ({ children }) => {
  const { client } = useContext(getApolloContext());
  const [rehydrated, setState] = useState(false);

  useEffect(() => {
    if (client instanceof AWSAppSyncClient) {
      (async () => {
        await client.hydrated();
        setState(true);
      })();
    }
  }, [client]);
  return rehydrated ? <>{children}</> : null;
};

export default Rehydrated;
```

**Comment by praisegeek** [comment]
This is only a temporary solution. Mutations hang on the client currently.

**Comment by TeoTN** [comment]
It won't help as it seems `AWSAppSyncClient` diverged too far from `ApolloClient`. It'll only work w/react-apollo v3 if you disable offline entirely using `disableOffline: true` in client's options

**Comment by amcdnl** [comment]
@TeoTN - I disabled offline and still got the issue. I bumped to `3.0.1` wonder if that has anything to do w/ it?

**Comment by TeoTN** [comment]
@amcdnl well, I forgot to mention that I'm using the Rehydrate from @dai-shi AND disabling offline. I'm pretty sure this is a dumb solution (the Rehydrate is not needed probably when offline's off) but it temporarily does the thing.

**Comment by amcdnl** [comment]
@TeoTN - Hmm, still didn't work for me. I'll just hold out for the official update (fingers crossed).

**Comment by ConnectivityChris** [comment]
> 
> 
> @TeoTN - Hmm, still didn't work for me. I'll just hold out for the official update (fingers crossed).

I'm using @dai-shi 's Rehydrate and adding implicit any to my client and seems to work
```
<ApolloProvider client={client as any}>
  <Rehydrated>
    <Router />
  </Rehydrated>
</ApolloProvider>
```

**Comment by doldyn** [comment]
I tried this as well last weekend. I got past the rehydrate by writing a separate one but the mutations don't process. They get added to the redux-offline queue but do not execute. AppSync has an offline-link implemented which is defining the effect. I suspect there are things that needs to be sorted at that level. For mutations, I was unable to make it work without fundamentally changing AppSync core. I hope we will see an official adaptation of this soon.

**Comment by SebSchwartz** [comment]
Can AWS tell us what is the plan for supporting Apollo 3.0? Is it in the roadmap? Is there an ETA? 
@elorzafe @manueliglesias @dabit3

**Comment by amcdnl** [comment]
Heres the complete solution I ended up with:

In my `package.json`, I have:

```json
{
  "dependencies": {
    "apollo-client": "^2.6.4",
    "apollo-link": "^1.2.12",
    "apollo-link-error": "^1.1.11",
    "aws-amplify": "^1.1.36",
    "aws-amplify-react": "^2.3.12",
    "aws-appsync": "^1.8.1",
    "react-apollo": "^3.0.1",
    "react": "^16.9.0",
    "react-dom": "^16.9.0",
    "aws-appsync-react": "^1.2.9",
  },
  "resolutions": {
    "apollo-client": "2.6.3"
  }
}
```

then I used one of the `Rehydrates` above which looks like:

```ts
import React, { useEffect, useState } from 'react';
import { useApolloClient } from 'react-apollo';
import AWSAppSyncClient from 'aws-appsync';

export const Rehydrated = ({ children }) => {
  const client = useApolloClient();
  const [rehydrated, setState] = useState(false);

  useEffect(() => {
    if (client instanceof AWSAppSyncClient) {
      (async () => {
        await client.hydrated();
        setState(true);
      })();
    }
  }, [client]);

  return rehydrated ? <>{children}</> : null;
};
````

then my AWS config looks like:

```ts
import Amplify from 'aws-amplify';
import { Auth } from 'core/Auth';
import AWSAppSyncClient, { createAppSyncLink } from 'aws-appsync';
import { ApolloLink } from 'apollo-link';
import { onError } from 'apollo-link-error';
import { AWS_CUSTOM_CONFIG } from '../../aws-custom-exports';

Amplify.configure(AWS_CUSTOM_CONFIG);

const onErrorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.map(({ message, locations, path }) =>
      console.error(
        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );

  if (networkError) {
    console.error(`[Network error]: ${networkError}`);
  }
});

const auth = {
  type: AWS_CUSTOM_CONFIG.aws_appsync_authenticationType as 'OPENID_CONNECT',
  jwtToken: () => {
    const session = Auth.getStoredSession();
    if (session) {
      return session.idToken;
    }
  }
};

export const appSyncClient = new AWSAppSyncClient(
  {
    url: AWS_CUSTOM_CONFIG.aws_appsync_graphqlEndpoint,
    disableOffline: true,
    region: AWS_CUSTOM_CONFIG.aws_appsync_region,
    auth
  },
  {
    link: ApolloLink.from([
      onErrorLink,
      createAppSyncLink({
        url: AWS_CUSTOM_CONFIG.aws_appsync_graphqlEndpoint,
        region: AWS_CUSTOM_CONFIG.aws_appsync_region,
        auth,
        complexObjectsCredentials: () => Auth.getStoredSession()
      })
    ])
  }
);
```

then I bring it all together in the `index.tsx` like:

```ts
import React, { Suspense, lazy } from 'react';
import ReactDOM from 'react-dom';
import { appSyncClient, Rehydrated } from 'core/Aws';
import { ApolloProvider } from 'react-apollo';

const Root = () => (
      <ApolloProvider client={appSyncClient as any}>
        <Rehydrated>
          Hello
        </Rehydrated>
      </ApolloProvider>
);

ReactDOM.render(<Root />, document.getElementById('root'));
````

**Comment by doldyn** [comment]
We have now explored this issue further. The reason to use AppSync over just the ApolloClient is if you need the offline capability. We wrote our own Rehydrated and got around the issues there but then when posting a mutation, it got enqueued in redux-offline but never processed. We debugged the flow and found the issues to be in the **offline-link** created redux-offline effect. 

AppSync is using a method called ```client.queryManager.buildOperationForLink``` that no longer exists to create the observable and hence nothing happens when redux-offline calls the effect. 

We refactored this to the following construct using the same approach as Apollo does in the mutations:
```
        client.queryManager.getObservableFromLink(
            mutation,
            extraContext,
            variables,
            false,
        ).subscribe({            
            next: data => {...
```


There were a few other minor issues that we have been changing to make the framework work for our specific use case but this is the key one the core framework falls over on. 

We now have this working nicely with hooks and our adapted version of AppSync for our use case and I just wanted to share in case anyone else are looking to do the same or if this could inspire the core team to update the framework as it is a hinderance for adaptation.

**Comment by sakhmedbayev** [comment]
We are also stuck here. We are using AppSync with Next.js. The solution found here[https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/450#issuecomment-539910312](url) works only client-side. The server-side cannot go beyong `Loading...` stage :-(

**Comment by undefobj** [comment]
Hello everyone - I want to reply and assure you that we have not abandoned support of this project. Per my earlier comment, this is a complex process and to be completely transparent there are things in the Apollo library that are out of our control and there have also been breaking changes in Apollo versions which we're trying to account for, which is also why we pinned the version. We've been spending many hours trying to account for this but it's not simple to solve. After putting more thought into this we've got some initial solutions that I hope will unblock many of you. What we've done is packaged two of the "Links" which are part of the SDK - the Auth and Subscription Links - and published them to NPM. This will allow you to include them in the newer Apollo versions if you wish to use the newer features. This does not include the "Offline Link" for offline support which is more complex to solve and we're looking into for the future.

Installation of the links:

```
npm install aws-appsync-auth-link aws-apsync-subscription-link
```

Usage:
https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support
 
With is I would then give the following recommendations for choosing your client strategy:

1. If you do **not** have offline use cases, you can either use the Auth & Subscription Links above with the latest Apollo client or alternatively use the Amplify GraphQL client instead of Apollo:https://aws-amplify.github.io/docs/js/api#amplify-graphql-client  
2. If you **do have offline use** cases please use the current version as-is with the older Apollo version. We're still working on a future strategy for the Offline Link.

Please let us know if you have any trouble using these links or if there is a scenario which we haven't accounted for here.

**Comment by SebSchwartz** [comment]
Thanks for giving us updates on this @undefobj .
For offline support, what about the solution of @doldyn ?

**Comment by undefobj** [comment]
> Thanks for giving us updates on this @undefobj .
> For offline support, what about the solution of @doldyn ?

The problem is query manager uses some private methods, and when you factor in the new Local State features of Apollo which can override the cache there are several scenarios which can lead to inconsistent state of the underlying persisted data. Essentially you have a race condition of multiple "writers" which is why we need a deterministic mechanism to update the cache, otherwise we cannot guarantee data integrity.

**Comment by tgjorgoski** [comment]
@undefobj ,
In the aws-appsync code there is also another links added when constructing the client:
        new link_1.ComplexObjectLink(complexObjectsCredentials) 
When I saw it, I guessed that this is needed for "complex object" support for AppSync where it also involves working with connected files on S3. Is that right?
I'm not using this functionality ATM, but would this functionality still be working with the new proposed solution with new Apollo with included auth and sub. links?

**Comment by undefobj** [comment]
> @undefobj ,
> In the aws-appsync code there is also another link added when constructing the client:
> ` new link_1.ComplexObjectLink(complexObjectsCredentials)`
> When I saw it, I guessed that this is needed for "complex object" support for AppSync where it also involves working with connected files on S3. Is that right?
> I'm not using this functionality ATM, but would this functionality still be working with the new proposed solution with new Apollo with included auth and sub. links?

Right now this is a separate link which is not published and still requires the older Apollo version. If there is high demand we can investigate publishing this as well, but I'd rather see if the core issue can be resolved.

**Comment by dai-shi** [comment]
Thanks to https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/448#issuecomment-541229659, it's now much cleaner with apollo-client. (without offline use cases)

Some notes:

1. a minor typo `apsync`.

2. it shows an error: `You are calling concat on a terminating link, which will have no effect`.
Modifying the code from the [doc](https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support)
```javascript
const link = ApolloLink.from([
  createAuthLink({ url, region, auth }),
  createSubscriptionHandshakeLink(url),
  createHttpLink({ uri: url }),
]);
```
to
```javascript
const link = ApolloLink.from([
  createAuthLink({ url, region, auth }),
  createSubscriptionHandshakeLink(url),
]);
```
the error message is gone.
I think this is fine because of
https://github.com/awslabs/aws-mobile-appsync-sdk-js/blob/50185bb0ff97191eae38e76d869552aad6ce09ad/packages/aws-appsync-subscription-link/src/index.ts#L33
and it has `resultsFetcherLink = createHttpLink({ uri: url }))` as default.

I wish somebody could confirm that.

**Comment by elorzafe** [comment]
@dai-shi thanks for pointing this out

**Comment by sakhmedbayev** [comment]
Using Typescript with (this)[https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support] example. Here:
 
```
  const auth = {
    type: AUTH_TYPE.AWS_IAM,
    credentials: () => Auth.currentCredentials()
  }

  const link = ApolloLink.from([
    createAuthLink({ url, region, auth }),
  ]);
```

 `auth` gets:

```
ype '{ type: AUTH_TYPE; credentials: () => Promise<ICredentials>; }' is not assignable to type 'AuthOptions'.
  Type '{ type: AUTH_TYPE; credentials: () => Promise<ICredentials>; }' is not assignable to type 'AuthOptionsIAM'.
    Types of property 'type' are incompatible.
      Type 'AUTH_TYPE' is not assignable to type '"AWS_IAM"'.ts(2322)
index.d.ts(5, 5): The expected type comes from property 'auth' which is declared here on type '{ url: string; region: string; auth: AuthOptions; }'
```

**Comment by dai-shi** [comment]
Putting ` as const` after `AWS_IAM`  might solve it.

**Comment by sakhmedbayev** [comment]
Thank @dai-shi! Facing another issue using [proposed](https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support) solution with Next.js. I posted a separate issue [here](https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/473).

**Comment by sakhmedbayev** [comment]
> Hello everyone - I want to reply and assure you that we have not abandoned support of this project. Per my earlier comment, this is a complex process and to be completely transparent there are things in the Apollo library that are out of our control and there have also been breaking changes in Apollo versions which we're trying to account for, which is also why we pinned the version. We've been spending many hours trying to account for this but it's not simple to solve. After putting more thought into this we've got some initial solutions that I hope will unblock many of you. What we've done is packaged two of the "Links" which are part of the SDK - the Auth and Subscription Links - and published them to NPM. This will allow you to include them in the newer Apollo versions if you wish to use the newer features. This does not include the "Offline Link" for offline support which is more complex to solve and we're looking into for the future.
> 
> Installation of the links:
> 
> ```
> npm install aws-appsync-auth-link aws-appsync-subscription-link
> ```
> 
> Usage:
> https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support
> 
> With is I would then give the following recommendations for choosing your client strategy:
> 
> 1. If you do **not** have offline use cases, you can either use the Auth & Subscription Links above with the latest Apollo client or alternatively use the Amplify GraphQL client instead of Apollo:https://aws-amplify.github.io/docs/js/api#amplify-graphql-client
> 2. If you **do have offline use** cases please use the current version as-is with the older Apollo version. We're still working on a future strategy for the Offline Link.
> 
> Please let us know if you have any trouble using these links or if there is a scenario which we haven't accounted for here.

@undefobj, does this solution works for SSR with IAM auth type? So far, I could not make it work. I am setting it in the following way:

```
const auth = {
        type: AUTH_TYPE.AWS_IAM as const,
        credentials: () => Auth.currentCredentials()
      }

 const link = ApolloLink.from([
    createAuthLink({ url, region, auth }),
    createHttpLink({ uri: url })
  ]);
```

and in CloudWatch I see this error:

```
Request Headers: {x-amzn-ErrorType=[UnauthorizedException]}
```

**Comment by undefobj** [comment]
@sakhmedbayev I can't see why it wouldn't, nothing specific here is different with a static build. What I have seen in the past is that when people host apps and put an extra DNS layer in front of AppSync with IAM the signature will get stripped since it contains region information.

**Comment by CodySwannGT** [comment]
[This solution](https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support) doesn't seem to work with any other auth strategy other than API Key.

**Comment by undefobj** [comment]
> [This solution](https://github.com/awslabs/aws-mobile-appsync-sdk-js#using-authorization-and-subscription-links-with-apollo-client-no-offline-support) doesn't seem to work with any other auth strategy other than API Key.

We've tested with React apps, however maybe this could be something with SSR. That's not something that has been tested.

**Comment by CodySwannGT** [comment]
I'm trying it in a react native expo app and didn't have any joy. Had to resort to the above Rehydrate hack, which is having other undesirable side effects.

Will keep going and report back with other findings.

**Comment by undefobj** [comment]
> I'm trying it in a react native expo app and didn't have any joy. Had to resort to the above Rehydrate hack, which is having other undesirable side effects.
> 
> Will keep going and report back with other findings.

Rehydrate is only used for Offline Support, which isn't supported in the custom links solution.

**Comment by CodySwannGT** [comment]
Got it. That worked. Have a new issue with sessions not persisting on reload, but I'll open a new issue with the amplify team

**Comment by 3nvi** [comment]
Worked great for my use case. In case anyone is interested in other authenticated methods, take a look at the definitions for the other auth methods in order to know what params to pass:

https://github.com/awslabs/aws-mobile-appsync-sdk-js/blob/68ba2b9862515f7ecba9bc8c333b9e716a0762bb/packages/aws-appsync-auth-link/src/auth-link.ts#L128

**Comment by sakhmedbayev** [comment]
https://github.com/awslabs/aws-mobile-appsync-sdk-js/blob/68ba2b9862515f7ecba9bc8c333b9e716a0762bb/packages/aws-appsync-auth-link/src/auth-link.ts#L153

Does it mean that Cognito User Pools is not yet supported???

**Comment by morrys** [comment]
Hi guys,
a few months ago I opened the issue https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/446 where I asked if it would be possible to evaluate an alternative to redux-offline to manage the offline link.

The libraries I created are compatible with react-relay 3.0 and I'm working on [apollo-client 3.0 support](https://github.com/morrys/wora/issues/25) (I'm also opening some PR in apollo-client for better integration https://github.com/apollographql/apollo-client/pull/5601).

Is it possible to have your feedback (even if negative)?

thanks

**Comment by Denakol** [comment]
> https://github.com/awslabs/aws-mobile-appsync-sdk-js/blob/68ba2b9862515f7ecba9bc8c333b9e716a0762bb/packages/aws-appsync-auth-link/src/auth-link.ts#L153
> 
> Does it mean that Cognito User Pools is not yet supported???

@sakhmedbayev 
 Cognito User Pools is support already
```
const auth = {
    type: awsConfig.aws_appsync_authenticationType,
    jwtToken: (async () =>  {
        const currentSession = await Auth.currentSession();
        return currentSession.getIdToken().getJwtToken()
    })
};
```

**Comment by VicFrolov** [comment]
Hello,

I got this working, but unfortunately for logged out users (using cognito pools), queries [that used to work] are perpetually stuck on 'loading', with no data for 'data' or 'error'. I have tried to rerun `amplify auth update`, and ensured that I selected unauthed users to have access: ` Allow unauthenticated logins? (Provides scoped down permissions that you can control via AWS IAM) Yes`.

In the console:

```
"Possible Unhandled Promise Rejection (id: 1):
"No current user"
```


`Auth.currentCredentials` returns what appears to be a  valid `CognitoIdentityCredentials` object, currentSession throws an error saying no current user.

I followed @undefobj 's writeup. 

```
const url = awsconfig.aws_appsync_graphqlEndpoint;
const region = awsconfig.aws_appsync_region;
const auth = {
  jwtToken: async () =>
    (await Auth.currentSession()).getIdToken().getJwtToken(),
  type: awsconfig.aws_appsync_authenticationType,
} as AuthOptions;

const httpLink = createHttpLink({ uri: url });
const link = ApolloLink.from([
  createAuthLink({ auth, region, url }),
  createSubscriptionHandshakeLink(url, httpLink),
]);

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link,
});
```

Any suggestions?

**Comment by undefobj** [comment]
> #446

@morrys I'm not sure about using something other than redux-offline. Looking at the alternatives, they don't address the core issue which is access to some of the cache update and notification properties in Apollo which is what we're struggling to solve in the newer versions. A couple team members are looking into this right now but it might take some efforts from the Apollo maintainers as well. We're evaluating a plan for how this could be offered as a custom link in 2020.

That being said there will always be fundamental struggles with doing offline when you're using a cache of query results which is what you have to do in the Apollo design. While we're going to continue to try and improve this for customers that want to use Apollo client we also have worked on an alternative solution with Amplify DataStore just released last week. If you're interested in offline apps it's our recommended solution to look first going forward: https://aws-amplify.github.io/docs/js/datastore

**Comment by undefobj** [comment]
> Hello,
> 
> I got this working, but unfortunately for logged out users (using cognito pools), queries [that used to work] are perpetually stuck on 'loading', with no data for 'data' or 'error'. I have tried to rerun `amplify auth update`, and ensured that I selected unauthed users to have access: ` Allow unauthenticated logins? (Provides scoped down permissions that you can control via AWS IAM) Yes`.
> 
> In the console:
> 
> ```
> "Possible Unhandled Promise Rejection (id: 1):
> "No current user"
> ```
> 
> `Auth.currentCredentials` returns what appears to be a valid `CognitoIdentityCredentials` object, currentSession throws an error saying no current user.
> 
> I followed @undefobj 's writeup.
> 
> ```
> const url = awsconfig.aws_appsync_graphqlEndpoint;
> const region = awsconfig.aws_appsync_region;
> const auth = {
>   jwtToken: async () =>
>     (await Auth.currentSession()).getIdToken().getJwtToken(),
>   type: awsconfig.aws_appsync_authenticationType,
> } as AuthOptions;
> 
> const httpLink = createHttpLink({ uri: url });
> const link = ApolloLink.from([
>   createAuthLink({ auth, region, url }),
>   createSubscriptionHandshakeLink(url, httpLink),
> ]);
> 
> const client = new ApolloClient({
>   cache: new InMemoryCache(),
>   link,
> });
> ```
> 
> Any suggestions?

@VicFrolov this looks to be more of an issue with your auth state from User Pools, which could be several things. I'd suggest moving this specific question into the [Amplify forums](https://github.com/aws-amplify/amplify-js/issues) for troubleshooting with a team member.

**Comment by morrys** [comment]
hi @undefobj,

> @morrys I'm not sure about using something other than redux-offline. Looking at the alternatives, they don't address the core issue which is access to some of the cache update and notification properties in Apollo which is what we're struggling to solve in the newer versions.
A couple team members are looking into this right now

What I'd like to have is a discussion with you about the alternative I propose. For those who know Apollo are two libraries ([wora/apollo-cache](https://github.com/morrys/wora/tree/master/packages/apollo-cache) & [wora/apollo-offline](https://github.com/morrys/wora/tree/master/packages/apollo-offline)) that do not require a huge effort.

> but it might take some efforts from the Apollo maintainers as well. We're evaluating a plan for how this could be offered as a custom link in 2020.

I agree, for this reason I propose PR to Apollo in order to improve integration.

> If you're interested in offline apps it's our recommended solution to look first going forward: https://aws-amplify.github.io/docs/js/datastore

I'll try it :)


Meanwhile, thanks for the answer and if I have not convinced you to evaluate my approach (also used to manage relay [react-relay-appsync](https://github.com/morrys/react-relay-appsync)), you can also close the issue https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/446

**Comment by undefobj** [comment]
@morrys I think we can leave it open for now. It's not that alternatives to redux offline aren't worth evaluating, it's more that isn't the current blocker that we're trying to resolve. Even if there was a different offline link implementation we'd still have the same core Apollo cache situation to overcome.

**Comment by morrys** [comment]
@undefobj, my approach to managing the offline is to extend the apollo client without using offline link.

The biggest problem in managing the apollo cache, in offline context, is caused by the use of server-side generated IDs. This problem is easily overcome by generating client-side IDs or avoiding "mutations" of new items.
(I think this is one of your problems)

I have overcome other cache iteration problems with wora/apollo-cache (more details are in this PR https://github.com/apollographql/apollo-client/pull/5601)

Another PR that will surely be interesting for you too is this: https://github.com/apollographql/apollo-client/pull/5649

Other problems can be caused during the parallel execution of mutations in the queue, default behavior of redux-offline, while In wora/apollo-offline the default execution is serial and is easily configurable for each mutation through its [functions of callback](https://morrys.github.io/wora/docs/apollo-offline#apollo-client-with-offline-options)

**Comment by undefobj** [comment]
Server generated IDs aren't an issue we have a request mapping for that which results in a deterministic update of the query references. The problems are more related to global notification of cache updates using the QueryManager as well as consistency issues reconciled with local resolvers.

**Comment by morrys** [comment]
I have not detected these problems .. do you have any issues with more details or some example project where the error is noticeable?

**Comment by sakhmedbayev** [comment]
@undefobj, are there good examples of how to use DataStore with either Next.js or Gatsby? I would greatly appreciate it if you can point me somewhere. Thanks!

**Comment by tugzera** [comment]
I have fixed the same error by downgrading the apollo to 2.5.8 .. But i still get some warninig: 

Possible Unhandled Promise Rejection (id: 0):
TypeError: undefined is not an object (evaluating 'netinfo_1.default.isConnected.fetch')

"dependencies": {
"@react-native-community/async-storage": "^1.7.1",
"@react-native-community/netinfo": "^5.0.0",
"amazon-cognito-identity-js": "^3.2.0",
"apollo-cache-hermes": "^0.8.10",
"aws-amplify": "^2.2.0",
"aws-amplify-react-native": "^2.2.3",
"aws-appsync": "^3.0.2",
"aws-appsync-auth-link": "^2.0.1",
"aws-appsync-react": "^3.0.2",
"aws-appsync-subscription-link": "^2.0.1",
"react": "16.9.0",
"react-apollo": "^2.5.8",
"react-native": "0.61.5",
"reactotron-react-native": "^4.0.2"
},

**Comment by jonixj** [comment]
Any progress on this issue? Also, publishing the complexObjectLink, compatible with the new version would be nice.

**Comment by TraMZzz** [comment]
I have the same issue

**Comment by kennu** [comment]
I ran into this issue trying to follow the basic instructions in the README which don't mention incompatibility with react-apollo 3.x. Perhaps they should clearly state that AppSync requires you to use an old Apollo version.

**Comment by VicFrolov** [comment]
Hi @undefobj , how would we go about implementing this with multiple auths. I am using IAM as a secondary auth for unauthenticated users.

```
const auth = {
        type: AUTH_TYPE.AWS_IAM,
        credentials: () => Auth.currentCredentials()
}

const unauth = {
  credentials: async () =>
    Auth.currentCredentials().then(value => ({
      accessKeyId: value.accessKeyId,
      secretAccessKey: value.secretAccessKey,
      sessionToken: value.sessionToken,
    })),
  type: AUTH_TYPE.AWS_IAM,
} as AuthOptions;

 const link = ApolloLink.from([
    createAuthLink({ url, region, auth }),
    createHttpLink({ uri: url })
  ]);
```

I'm not sure what to do next. I've tried adding a second `createAuthLink` where I also have `auth: unauth`, but that wasn't working properly.

Any help is much appreciated!

**Comment by danielbayerlein** [comment]
Thanks for the solution @undefobj. I wrote a blog post about it. Hope it helps the one or the other.
https://dev.to/danielbayerlein/migrate-apollo-from-v2-to-v3-in-conjunction-with-aws-appsync-16c0

**Comment by willsamu** [comment]
Hello everyone, I have just [published an article](https://dev.to/willsamu/how-to-get-aws-appsync-running-with-offline-support-and-react-hooks-678) on how to set up AWS AppSync with react apollo 3.x, react hooks **and** offline support. Hope it helps!

**Comment by amcdnl** [comment]
Another good library I found for this: https://github.com/wolfeidau/appsync-apollo-links

**Comment by NTonani** [comment]
ETA around 3.0 support, including offline?

**Comment by BenoitDel** [comment]
Appsync is now two major release behind for its apollo/client dependency. the solution given by @undefobj works but makes the project depends both on apollo 2.4 and apollo 3.0 which has major impact on bundle size.
Now that apollo 3.0 has reach stable version i hope we will see progress. Any updates ?

**Comment by zuhair-naqvi** [comment]
Can someone please confirm if this is at least on the Road Map for future releases?

**Comment by dylan-westbury** [comment]
This solution worked for me, react native with hooks. 

**App.js**

```
import React from 'react';
import { Provider } from 'react-redux';
import { ApolloProvider } from '@apollo/client';
import { store } from './store';
import Navigation from './Navigation';
import theme from './components/Theme';
import { ThemeProvider } from 'react-native-elements';
import Rehydrated from './components/Rehydrated';
import AWSAppSyncClient from 'aws-appsync';
import awsmobile from './aws-exports';
import Amplify, { Auth } from 'aws-amplify';

Amplify.configure({
  ...awsmobile,
  PushNotification: {
    requestIOSPermissions: false,
  },
});

const client = new AWSAppSyncClient({
  url: awsmobile.aws_appsync_graphqlEndpoint,
  region: awsmobile.aws_appsync_region,
  auth: {
    type: awsmobile.aws_appsync_authenticationType,
    jwtToken: async () =>
      ( await Auth.currentSession() ).getIdToken().getJwtToken(),
  },
  disableOffline: true,
});

const App = () => {
  return (
    <Provider store={ store }>
      <ApolloProvider client={ client }>
        <Rehydrated client={ client }>
            <ThemeProvider theme={ theme }>
              <Navigation/>
            </ThemeProvider>
        </Rehydrated>
      </ApolloProvider>
    </Provider>
  );
};

export default App;

```

**Rehydrated component**
```
import React, { useContext, useEffect, useState } from 'react';
import { getApolloContext } from '@apollo/react-hooks';
import AWSAppSyncClient from 'aws-appsync';

const Rehydrated = ({ children }) => {
  const { client } = useContext(getApolloContext());
  const [ rehydrated, setRehydrated ] = useState(false);

  useEffect(() => {
    if (client instanceof AWSAppSyncClient) {
      ( async () => {
        await client.hydrated();
        setRehydrated(true);
      } )();
    }
  }, [ client ]);
  return rehydrated ? <>{ children }</> : null;
};

export default Rehydrated;

```

**Comment by Neuromancer93** [comment]
@dylan-westbury  I think the solution you posted above works with react native but what most of us are looking for here is a solution that works with the current version of apollo and maintains the offline functionality which looks like you had to disable in your example. The people want both!

**Comment by ablbol** [comment]
Please add support to complexObjectsCredentials with latest Apollo version 3.0.  I really need it for my work.
Thanks

**Comment by AstroneCommons** [comment]
@ablbol This issue is opened for more than a year now. This library is barely maintained (alive). So good luck

**Comment by manueliglesias** [comment]
Hi,

Just to let you know, we've published new versions of the links to make them compatible with Apollo 3 (no offline support)

For more details: https://github.com/awslabs/aws-mobile-appsync-sdk-js/pull/561#issuecomment-701696316

**Comment by andrekovac** [comment]
@manueliglesias @undefobj @elorzafe Any timeline to upgrade `aws-appsync-react` to support `@apollo/client` version 3 with **offline support** ?

The offline functionality is the main reason why I go for the AWS/Apollo stack

**Comment by andrekovac** [comment]
Another question: Could https://github.com/apollographql/apollo-cache-persist bring offline support together with the packages @manueliglesias published to support Apollo 3?

**Comment by andrekovac** [comment]
@tgjorgoski Was your thumbs-up emoji confirmation that it does work with `apollo-cache-persist`? Or do you have the same question as I do?

**Comment by tgjorgoski** [comment]
@andrekovac  , just liked the question as an interesting one, I would also like to know.

**Comment by aagamdoshi** [comment]
> @TeoTN - Hmm, still didn't work for me. I'll just hold out for the official update (fingers crossed).

Did you get any luck on this?

**Comment by fab-mindflow** [comment]
> @manueliglesias @undefobj @elorzafe Any timeline to upgrade `aws-appsync-react` to support `@apollo/client` version 3 with **offline support** ?
> 
> The offline functionality is the main reason why I go for the AWS/Apollo stack

Any update please?

**Comment by amcdnl** [comment]
@fab-mindflow - I'll let you know I gave up on this - its actually way cleaner just to do just apollo. Also the libraries require AWS-SDK which creates a huge bloat and its not tree shakeable.

**Comment by LYevhen** [comment]
@undefobj That is very sad that after two years there is no progress, roadmaps and even hope for others to have this important update.

**Comment by tabdon** [comment]
@amcdnl Could you please elaborate on your setup? I'd like to ditch AWS-SDK as well. Thanks

**Comment by razor-x** [comment]
Here is what I have been using in production for a while now:

➡️ https://gist.github.com/razor-x/e19d7d776cdf58d04af1e223b0757064

Versions:

```
{
  "@apollo/client": "^3.3.12",
  "apollo-cache-persist": "^0.1.1",
  "graphql": "^15.0.0",
  "subscriptions-transport-ws": "^0.9.16"
}
```

The WebSocket integration is based off of https://gist.github.com/zachboyd/f5630736b0a5a9b627d61bfd25299c90 and the custom domain support is from https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/517#issuecomment-786772241

Only issue I've noticed is that I'm not 100% the subscriptions are as stable or reliable as I would expect. If anyone uses this and can share their findings, let me know.

**Comment by PatrykMilewski** [comment]
How it's possible that it's still not done after 3 years?

**Comment by shamseer-ahammed** [comment]
its 2023 nov 31 and still , no official fix for this
